contracts/src/Teleporter/upgrades/TeleporterRegistry.sol

```diff
@@ -5,28 +5,110 @@
 pragma solidity 0.8.18;
 
-import {WarpProtocolRegistry, ProtocolRegistryEntry} from "../../WarpProtocolRegistry.sol";
 import {ITeleporterMessenger} from "../ITeleporterMessenger.sol";
+import {IWarpMessenger, WarpMessage} from "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
 
 /**
- * @dev TeleporterRegistry contract is a {WarpProtocolRegistry} and provides an upgrade
- * mechanism for {ITeleporterMessenger} contracts.
+ * @dev Registry entry that represents a mapping between protocolAddress and version.
  */
-contract TeleporterRegistry is WarpProtocolRegistry {
-    constructor(
-        ProtocolRegistryEntry[] memory initialEntries
-    ) WarpProtocolRegistry(initialEntries) {}
+struct ProtocolRegistryEntry {
+    uint256 version;
+    address protocolAddress;
+}
+
+/**
+ * @dev TeleporterRegistry contract provides an upgrade mechanism for {ITeleporterMessenger} contracts
+ * through Warp off-chain messages
+ */
+contract TeleporterRegistry {
+    // Address that the off-chain Warp message sets as the "source" address.
+    // The address is not owned by any EOA or smart contract account, so it
+    // cannot possibly be the source address of any other Warp message emitted by the VM.
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    IWarpMessenger public constant WARP_MESSENGER =
+        IWarpMessenger(0x0200000000000000000000000000000000000005);
+
+    bytes32 public immutable blockchainID;
+
+    // The latest protocol version. 0 means no protocol version has been added, and isn't a valid version.
+    uint256 public latestVersion;
+
+    // Mappings that keep track of the protocol version and corresponding contract address.
+    mapping(uint256 version => address protocolAddress)
+        private _versionToAddress;
+    mapping(address protocolAddress => uint256 version)
+        private _addressToVersion;
 
     /**
-     * @dev Gets the {ITeleporterMessenger} contract of the given `version`.
+     * @dev Emitted when a new protocol version is added to the registry.
+     */
+    event AddProtocolVersion(
+        uint256 indexed version,
+        address indexed protocolAddress
+    );
+
+    /**
+     * @dev Emitted when the latest version is updated.
+     */
+    event LatestVersionUpdated(
+        uint256 indexed oldVersion,
+        uint256 indexed newVersion
+    );
+
+    /**
+     * @dev Initializes the contract by setting `blockchainID` and `latestVersion`.
+     * Also adds the initial protocol versions to the registry.
+     */
+    constructor(ProtocolRegistryEntry[] memory initialEntries) {
+        blockchainID = WARP_MESSENGER.getBlockchainID();
+        latestVersion = 0;
+
+        for (uint256 i = 0; i < initialEntries.length; i++) {
+            _addToRegistry(initialEntries[i]);
+        }
+    }
+
+    /**
+     * @dev Gets and verifies a Warp off-chain message, and adds the new version to procotol address mapping,
+     * specified in the Warp off-chain message, to the registry.
```

Comment:
Receives a Warp off-
```suggestion
     * @dev Receives a Warp off-chain message containing a new protocol version and address to be registered,
     * and adds the the new values to the respective mappings.
```

Technically the verification of the message happens in the precompile, not here :)